### PR TITLE
Fix validation of parametrized evented objects

### PR DESCRIPTION
**Pull request by brisvag**
# Description
This is an alternatively, and arguably better, implementation of #4609.

It's a lot cleaner because it actually overrides the `ModelField` machinery to act more cleanly, rather than doing weird stuff at the `Model` level. Most importantly, this does not force us to implement complicated sub_field validation logic for each class (`EventedList` and co) like we had to do in #4609.

cc @tlambert03.

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: napari/components/interaction_box.py [hunk 1]
```diff
@@ -35,18 +37,18 @@ class InteractionBox(EventedModel):
         Bool indicating whether the interaction box allows creating a new selection by dragging outside an existing interaction_box.
     """
 
-    points: Array[float, (-1, 2)] = None
+    points: Optional[Array[float, (-1, 2)]] = None
     show: bool = False
```
Comment (brisvag) [inline napari/components/interaction_box.py:41]: One benefit of this PR is that now we are forced to use the correct annotation in places like here :P

File: napari/utils/events/evented_model.py [hunk 2]
```diff
@@ -55,15 +71,112 @@ def no_class_attributes():
...
+
+
+@contextmanager
+def parametrized_generic_fix():
+    """
+    Context in which pydantic more consistently validates parametrized generics.
+
+    This context temporarily replaces ModelField with
+    ParametrizedGenericCompliantModelField.
+    It also permanently overrides pydantic.sequence_like() to be more lax.
```
Comment (brisvag) [inline napari/utils/events/evented_model.py:169]: This is an intentional side effect: without permanently replacing `pydantic`'s `sequence_like()` with a more lenient version, many types will simply fail when parametrized (see the end of [this issue in pydantic](https://github.com/samuelcolvin/pydantic/issues/4161) for an example)

**Comment by brisvag** [comment]
Ok, I figured out what's the issue with the Selection test error; it's a bit tricky. Basically, this is how validation occurs in pydantic:

1. validate pre-validators (which includes class validators defined with `__get_validators__`)
2. validate subfields (for example ensure all `int` elements in a `List[int]`)
3. validate post-validators (which include the `coerce_type` this PR adds)

Importantly, step 2. ends with these lines in pydantic:

https://github.com/samuelcolvin/pydantic/blob/f529e0d3541d3cab0aa9c2795a747cd8af4968e5/pydantic/fields.py#L912-L929

So in the case of `Selection`, this happens:
1. selection is validated by the custom validator, which returns a `Selection` object with a correctly initialized value for `sel._current`
2. subfields are validated, then returned as a basic `set`. This forgets completely of the `_current` value
3. `coerce_type` returns a `Selection`, now with an unitialized `_current`, since it was created with a simple set as input.

I'm trying to figure out if I can override this...

File: napari/utils/events/containers/_selection.py [hunk 3]
```diff
@@ -148,40 +144,18 @@ def validate(cls, v, field: 'ModelField'):
             data = v
             current = None
 
-        if not sequence_like(data):
+        if not isinstance(data, Iterable):
```
Comment (tlambert03) [inline napari/utils/events/containers/_selection.py:147]: one of the main downsides of `Iterable` is that it also supports `str`... would that work here?

```py
In [1]: from pydantic.utils import sequence_like

In [2]: sequence_like('asdf')
Out[2]: False

In [3]: isinstance('asfadsf', Iterable)
Out[3]: True
```

File: napari/utils/events/evented_model.py [hunk 4]
```diff
@@ -1,11 +1,27 @@
...
+    get_origin,
+)
 
 import numpy as np
-from pydantic import BaseModel, PrivateAttr, main, utils
+import pydantic
+from pydantic import BaseModel, PrivateAttr
+from pydantic import main as pydantic_main
+from pydantic import utils as pydantic_utils
+from pydantic.fields import SHAPE_SET, ModelField, Validator, prep_validators
```
Comment (tlambert03) [inline napari/utils/events/evented_model.py:23]: maybe also use `pydantic_fields` to make it more obvious below?

File: napari/utils/events/evented_model.py [hunk 5]
```diff
@@ -55,15 +71,112 @@ def no_class_attributes():
...
+            for f in self.validators
+            if f.__name__ != 'arbitrary_type_validator'
+        ]
+
+        origin = get_origin(self.outer_type_) or self.outer_type_
+        # anything that's not really a class should not be used for coercion
+        if (
+            not isinstance(origin, type)
+            or origin is Union
+            or issubclass(origin, Protocol)
```
Comment (tlambert03) [inline napari/utils/events/evented_model.py]: if `origin is Union` then isn't it by definition not an instance of `type`? 
```python
In [20]: isinstance(Union, type)
Out[20]: False
```

also, why can't a Protocol subclass be used for coercion?

```python
In [21]: class P(Protocol):
    ...:     def x(self): ...
    ...:

In [22]: class C(P):
    ...:     def x(self): return 1
    ...:

In [23]: C()  # we could certainly call this origin and return an instance
Out[23]: <__main__.C object at 0x13db53f10>

In [24]: issubclass(C, Protocol)
Out[24]: True
```

The difference between `P` and `C` here is:
```python
In [27]: C._is_protocol
Out[27]: False

In [28]: P._is_protocol
Out[28]: True
```

File: napari/utils/events/evented_model.py [hunk 6]
```diff
@@ -55,15 +71,112 @@ def no_class_attributes():
...
+
+    This context temporarily replaces ModelField with
+    ParametrizedGenericCompliantModelField.
+    It also permanently overrides pydantic.sequence_like() to be more lax.
+    """
+    pydantic_main.ModelField = ParametrizedGenericCompliantModelField
+    seq_like = pydantic_utils.sequence_like
+
+    def sequence_like(v):
+        return seq_like or isinstance(v, Sequence)
```
Comment (tlambert03) [inline napari/utils/events/evented_model.py]: `and not isinstance(v, str)`?

File: napari/utils/events/evented_model.py [hunk 7]
```diff
@@ -79,21 +192,24 @@ class EventedMetaclass(main.ModelMetaclass):
 
     def __new__(mcs, name, bases, namespace, **kwargs):
         with no_class_attributes():
-            cls = super().__new__(mcs, name, bases, namespace, **kwargs)
+            with parametrized_generic_fix():
+                cls = super().__new__(mcs, name, bases, namespace, **kwargs)
```
Comment (tlambert03) [inline napari/utils/events/evented_model.py]:
```suggestion
        with no_class_attributes(), parametrized_generic_fix():
            cls = super().__new__(mcs, name, bases, namespace, **kwargs)
```
can use two context managers with a comma

**Review by tlambert03 — COMMENTED** [review]
Love it.  Looks fantastic.  I'm going to stop short of a full approval cause I'm a bit curious whether the test failures are hinting at a side effect here that maybe isn't immediately obvious?  But overall i think this is a stellar PR.  thanks @brisvag!

File: napari/utils/events/evented_model.py (hunk 5 above)
Comment (brisvag) [reply napari/utils/events/evented_model.py]: The problem here is `issubclass(A, B)` raising an exception if `A` is not an instance of `type`. I mixed a few checks together so this became a bit confusing (and maybe wrong). The `Protocol` check was added after I had some issue, but maybe those are solved now... wil check!

Comment (brisvag) [reply napari/utils/events/evented_model.py]: Renmoving protocol will fail with `Colormap`... If you can check that out it would be great! Just remove that line and run `pytest` :P

Comment (tlambert03) [reply napari/utils/events/evented_model.py]: > The problem here is issubclass(A, B) raising an exception if A is not an instance of type

right, you definitely need the `isinstance(origin, type)` check... but I'm saying the `is Union` check should be removed (it will always be False if origin is an instance of `type`), and the protocol check could be `getattr(origin, '_is_protocol', False)`

**Comment by brisvag** [comment]
Yes! I figured out the big remaining issue! We need to use the tricks described [in the numpy subclassing docs](https://numpy.org/doc/stable/user/basics.subclassing.html#slightly-more-realistic-example-attribute-added-to-existing-array) for every type that subclasses `np.ndarray`. Without this `__new__`, our `coerce_type` validator was simply passing the array to the *constructor* of `np.ndarray`, which is a complicated thing, very different from `np.array` :P

**Comment by brisvag** [comment]
Not sure what those `yaml` errors are. Any tips?

**Comment by Czaki** [comment]
> Not sure what those `yaml` errors are. Any tips?

check type of this 1 (using debugger and breakpoint on RepresenterError). It may be some numpy.int16 `1` instead of pure python `1`

File: napari/utils/color.py [hunk 8]
```diff
@@ -15,6 +15,11 @@ class ColorValue(np.ndarray):
     use the ``validate`` method to coerce a value to a single color.
     """
 
+    def __new__(cls, input_array):
+        # see https://numpy.org/doc/stable/user/basics.subclassing.html#slightly-more-realistic-example-attribute-added-to-existing-array
+        obj = np.asarray(input_array).view(cls)
```
Comment (andy-sweet) [inline napari/utils/color.py]: I considered this in [PR 4704](https://github.com/napari/napari/pull/4704#discussion_r900289069), though was considering something slightly fancier. Agreed that we should either do something like this or remove the `np.ndarray` as a sub-type (to indicate that this is just a pydantic field with some specific coercion logic) and I prefer doing this (probably because it plays nice with type checking in pydantic and in general).

One suggestion is actually to call `validate` instead of `asarray` here and in the other similar classes.

```suggestion
        obj = cls.validate(input_array).view(cls)
```

That's because I'd expect something like `ColorValue([0, 1])` to fail and maybe something like `ColorValue('red')` to do the coercion.

I guess this PR is an improvement as we can actually use the constructor more easily now, but doesn't quite feel complete.

Comment (andy-sweet) [reply napari/utils/color.py]: Or will that cause `validate` to be called twice now? If so should we consider removing `__get_validators__`?

Comment (brisvag) [reply napari/utils/color.py]: Actually no, it wouldn't be called twice because of this check in `coerce_type`:

```py
if isinstance(value, origin):
	return value
```

**Comment by brisvag** [comment]
> check type of this 1 (using debugger and breakpoint on RepresenterError). It may be some numpy.int16 1 instead of pure python 1

Mhm, it turns out it's a `ConstrainedIntValue` from our custom types... ~No clue why this is happening in this PR O.o~ Makes sense, it's getting coerced now, but it wasn't before. I added it to the yaml dumper, but maybe we prefer not to coerce this, and add some sort of exception?

**Comment by brisvag** [comment]
> So in the case of `Selection`, this happens:
> 
>     1. selection is validated by the custom validator, which returns a `Selection` object with a correctly initialized value for `sel._current`
> 
>     2. subfields are validated, then returned as a basic `set`. This forgets completely of the `_current` value
> 
>     3. `coerce_type` returns a `Selection`, now with an unitialized `_current`, since it was created with a simple set as input.

It turns out, the test was not checking this; instead, I had removed the validation for `_current`, which was present before. Now that I readded this, the test passes again.

However, this means that the above problem is still present.

**Comment by brisvag** [comment]
I added a little ad hoc snippet for `Selection` in the `Field` in order to fix the selection issue... Not great, but it should work.

**Comment by brisvag** [comment]
Ok, so this is a bit of a pickle: we want some subclasses or builtin or third party types to actually *be* coerced (such as `ColorArray`), but some of them are causing a bunch of issues when passed to external things: the previous failure in the `yaml` dumper was caused by passing to it our `ConstrainedInt`, and now more problems come up with things such as `settings._fields.Language` (which is just a string with validator) being passed to `qt` widgets which don't know what to do with it.

Maybe the solution is to re-add something I had earlier: a way to opt out of this coercion mechanism.

File: napari/settings/_appearance.py [hunk 9]
```diff
@@ -13,14 +16,12 @@ class AppearanceSettings(EventedModel):
         description=trans._("Select the user interface theme."),
         env="napari_theme",
     )
-    highlight_thickness: int = Field(
+    highlight_thickness: HighlightThickness = Field(
```
Comment (brisvag) [inline napari/settings/_appearance.py]: Without this change, the type of this field will be `pydantic.types.ConstrainedInt`, which obviously does not opt out of `_coerce_type`; this results in the value of this field being coerced by our logic, which causes all sorts of issues downstream (`yaml dumper` does not know what to do and `qt` widgets as well).

This problem will show up with any other validator-only types that pydantic generates under the hood. In future cases, we'll have to explicitly make them our subtypes that opt out of type coercion.

File: napari/settings/_fields.py [hunk 10]
```diff
@@ -12,6 +12,8 @@ class Theme(str):
     Custom theme type to dynamically load all installed themes.
     """
 
+    _coerce_type = False
+
```
Comment (brisvag) [inline napari/settings/_fields.py]: These classes are not real types that we care about; their only purpose is to provide validators. For these things, we want to opt out of type coercion.

File: napari/utils/color.py [hunk 11]
```diff
@@ -67,7 +71,7 @@ def validate(
         >>> ColorValue.validate('#ff0000')
         array([1., 0., 0., 1.], dtype=float32)
         """
-        return transform_color(value)[0]
+        return transform_color(value)[0].view(cls)
```
Comment (brisvag) [inline napari/utils/color.py:74]: I moved this `view(cls)` logic here so that calling `validate()` manually will return the proper class.

File: napari/utils/colormaps/standardize_color.py [hunk 12]
```diff
@@ -62,7 +62,10 @@ def transform_color(colors: Any) -> np.ndarray:
         invalid inputs
     """
     colortype = type(colors)
-    return _color_switch[colortype](colors)
+    for typ, handler in _color_switch.items():
+        if issubclass(colortype, typ):
+            return handler(colors)
+    raise ValueError(f"cannot convert type '{colortype}' to a color array.")
```
Comment (brisvag) [inline napari/utils/colormaps/standardize_color.py:68]: this was failing with subclasses such as `ColorValue`.

**Review by brisvag — COMMENTED** [review]
I added a few comments below to help review. I think there is one last wrinkle to smooth out: do we want to opt *out* of type coercion (as implemented now) - which makes everything work more magically but can cause surprises if new "fake types" are created/used and people are unaware - or opt *in*, which means we have to explicilty add `_coerce_type == True` to a bunch of stuff?

File: napari/utils/events/evented_model.py [hunk 13]
```diff
@@ -55,15 +71,118 @@ def no_class_attributes():
...
+            for f in self.validators
+            if f.__name__ != 'arbitrary_type_validator'
+        ]
+
+        origin = get_origin(self.outer_type_) or self.outer_type_
+        # anything that's not really a class should not be used for coercion
+        if (
+            not isinstance(origin, type)
+            or getattr(origin, '_is_protocol', False)
+            or not getattr(origin, '_coerce_type', True)
```
Comment (tlambert03) [inline napari/utils/events/evented_model.py]: I don't love this construct.  I just pulled and played with it a bit.  I started by commenting it out in all the settings fields ... but no tests failed (so that's one note... it would be easy for someone to remove them without knowing)

then I removed on `Array` to see the fail... the stack trace was extremely hard to parse... had I not been specifically reviewing this PR, I'm not sure I would have had any idea that the "solution" would have been to add back `_coerce_type` or indeed that it even had anything to do with logic added in this class at all.

I think the addition of _coerce type in `conint` also has a code smell. (hard to figure out why this would have any bearing on a very minor extension of a builtin pydantic object).

this just generally feels like something that "got this PR to work" ... but doesn't have the feel of a framework that napari developers will readily understand and employ correctly.  

It also feels like it overlaps a bit (or specifically opts out of) with the intention of pydantic in the first place.  I'm sure there's a good reason, but it's not immediately clear what this opting out here couldn't happen in the validator method of the type (rather than our own additional API on top of pydantic).  would love to get some more background/explanation on this and consider alternatives

Comment (brisvag) [reply napari/utils/events/evented_model.py]: > I started by commenting it out in all the settings fields ... but no tests failed

Some tests should definitely fail, that's why I needed to add it. Specifically, the `yaml` dumping of the settings failed because `Language` is not coerced to a normal `str`, and the dumper does not know what to do with it.

> then I removed on Array to see the fail... the stack trace was extremely hard to parse... 

Yes, this is just a general gripe of pydantic; the validation errors are super hard to chase down.

> I'm not sure I would have had any idea that the "solution" would have been to add back _coerce_type or indeed that it even had anything to do with logic added in this class at all.
>
> I think the addition of _coerce type in conint also has a code smell. (hard to figure out why this would have any bearing on a very minor extension of a builtin pydantic object).

I generally agree, and the solution would be to switch from an opt-out to an opt-in, but the consequence is that we lose coercion anything that is, for example, a `Sequence` that we didn't specifically subclass for this purpose. But maybe it's not that much work, and I can try to implement it that way to see what breaks down. (with the caveat that anything created by a user will also have to opt in, if that ever comes up).

> It also feels like it overlaps a bit (or specifically opts out of) with the intention of pydantic in the first place.

I disagree with this: pydantic is simply inconsistent with type coercion (subclasses of `str` just become `str`, some things return different types depending on their parametrization, other things are simply never coerced). This PR is essentially unifying this behaviour (or trying to), but there are unintended consequences because we are using *types* as a way to re-use validators across models, rather than simply having validators on each model.

> it's not immediately clear what this opting out here couldn't happen in the validator method of the type (rather than our own additional API on top of pydantic).

Because pydantic (sometimes, inconsistently) adds its own validators at the end of user-made ones, which override whatever type coercion we did.

Comment (tlambert03) [reply napari/utils/events/evented_model.py]: Is this additional coerce type thing mandatory for the original goal of this PR (to fix parametrized generics)?  Could it be considered on its own?  Or are they inextricably linked?

Comment (tlambert03) [reply napari/utils/events/evented_model.py]: > Some tests should definitely fail, that's why I needed to add it. Specifically, the yaml dumping of the settings failed because Language is not coerced to a normal str, and the dumper does not know what to do with it.

🤷‍♂️ ... they didn't fail.  perhaps there's something being skipped on macos, or required CI, or something

Comment (brisvag) [reply napari/utils/events/evented_model.py]: > Is this additional coerce type thing mandatory for the original goal of this PR (to fix parametrized generics)? Could it be considered on its own? Or are they inextricably linked?

They are linked: the problem with parametrized generics is that they fall in that category of things that get coerced to a "base" type (so no matter what you do in you `Sequence` subclass, it will always end up being a `list` if you parametrize it, and your validators will be ignored).

> man_shrugging ... they didn't fail. perhaps there's something being skipped on macos, or required CI, or something

:D bizarre! Anyways, it's not *always strictly necessary* to add `_coerce_type = False`. It only is if that value is then passed around to places where having, for example, a `class Language(str)` causes issues.

Comment (brisvag) [reply napari/utils/events/evented_model.py]: I just pushed a change to swap the logic: it's now opt-in. Notice that while I had to add the coercion to some things (Evented objects) it's not actually necessary everywhere: `ColorArray` is fine without it because it has coercion logic in its validator AND it's one of those cases where `class_validators` are respected.

Comment (tlambert03) [reply napari/utils/events/evented_model.py]: thanks, I'm having a play with `populate_validators` myself right now.  need to understand it a bit better before I can say anything intelligent :joy:

**Comment by tlambert03** [comment]
this really is a challenging problem, and exposes some very deep brokenness with pydantic's handling of generics.  I also find it very interesting :joy:, and some of the stuff I've been working on over in beartype ([with the TypeHint class](https://github.com/beartype/beartype/blob/main/beartype/door/_doorcls.py)) has taught me more about how to work with generics and parametrized generics.

I think the biggest problem we face here is `ModelField.shape`... and the way that pydantic choses that shape in `type_analysis` _and_ the degree to which it depends on that shape for many other things.  For example, it's a big problem that `self.shape` of a field for a parametrized generic (like `EventedList[float]`) is `SHAPE_SEQUENCE`  (as I'm sure you've become intimately aware 😂) and, to a lesser degree, it's also a problem that an unparameterized generic like `EventedList` is SHAPE_SINGLETON.  My first feeling here is that, if we're going to "go there" and create a `ParametrizedGenericCompliantModelField`, then we need to Go There™️ and create a new SHAPE altogether.

A couple things that might be helpful...

given a generic class:
```
from typing import Generic, TypeVar
T = TypeVar("T")
class MyGeneric(Generic[T]):
    ...
```

- you can tell whether the thing is a generic class by asking `issubclass(MyGeneric, Generic)`
- you can tell whether there are "free" parameters remaining by looking at `__parameters__`:
```python
In [24]: EventedList.__parameters__
Out[24]: (~_T,)

In [25]: EventedList[int].__parameters__
Out[25]: ()
```
(not immediately sure if that last thing is helpful, but it might be).


There's of course big range of what an end-user might do with generics.  the number of parameters could vary (i.e. mappings have two free params and sequences have 1 free param, etc...) and I think that's what pydantic was trying to capture with the whole `SHAPE` concept.  By not providing a hook for generics to offer their own validation, I think pydantic was doing a bit too much magic.  And so, in this `ParametrizedGenericCompliantModelField` class, we might consider just fully opting out of that magic, and either allowing generics to give their own validators, or making a new SHAPE concept that _only_ worries about the number of free parameters in the generic.  That is rather than SHAPE_SEQUENCE, SHAPE_MAPPING, etc... you would want to just track how many parameters the origin accepts (SHAPE_1PARAM, SHAPE_2PARAM, etc...)

note that for the purpose of type analysis at runtime, we can consider `TypeVar("T")` to be the same as `Any`... which means, Given a typehint of `hint = MyGeneric`, you can look at `hint.__parameters__`, see the unbound type var, and treat that as `MyGeneric[Any]`, whereas `MyGeneric[int]` has no free `__parameters__` (but _does_ have args):

```python
In [33]: MyGeneric.__parameters__
Out[33]: (~T,)

In [34]: get_args(MyGeneric)
Out[34]: ()

In [35]: MyGeneric[int].__parameters__
Out[35]: ()

In [36]: get_args(MyGeneric[int])
Out[36]: (<class 'int'>,)
```


I'm not sure any of that is helpful yet... but I just wanted to get those thoughts down. I have to do other things at the  moment, will come back to this later

**Comment by tlambert03** [comment]
sidenote:  `Array`, really is a generic (by nature of having a `__class_getitem__`)... so it should also probably have `Generic` added to the bases (unless that breaks due to some metaclass conflict...)

**Comment by tlambert03** [comment]
one more thought:  the goal of making pydantic work with parametrized generics is well-contained enough, and big enough, that you might consider extracting this (perhaps just temporarily) to it's own package (`pydantic-generic`), and its own set of tests.  We could just subclass `pydantic_generic.BaseModel` rather than `pydantic.BasModel` in `EventedModel`

**Comment by tlambert03** [comment]
aaand one more thought :)

we could take this in steps, and just try to solve `test_validate_evented_collections` first, and then `test_validate_parametrized_evented_collections` in a followup.

the following appears sufficient to solve the first test:
```python
def _make_arbitrary_type_validator(type_: Type[T]) -> Callable[[T], T]:
    def _arbitrary_type_validator(v: Any) -> T:
        if isinstance(v, type_):
            return v
        try:
            return type_(v)
        except Exception as e:
            raise errors.ArbitraryTypeError(expected_arbitrary_type=type_) from e
    return _arbitrary_type_validator

class ParametrizedGenericCompliantModelField(ModelField):
    ...
    def populate_validators(self):
        original_arb_validator = pydantic_validators.make_arbitrary_type_validator
        pydantic_validators.make_arbitrary_type_validator = _make_arbitrary_type_validator
        
        try:
            super().populate_validators()
        finally:
            pydantic_validators.make_arbitrary_type_validator = original_arb_validator
```

it doesn't fix the second test (`EventedList[float]`), because SHAPE then becomes SHAPE_SEQUENCE and that validator is no longer used


**edit**
furthermore, another argument for taking this in two steps is that the fix proposed here for parameterized `EventedList[float]` does work on instantiation, but it adds a new "problem" in that it doesn't convert to `EventedList(basetype=float)`  (yes, that's *also* "just" a pydantic problem ... but it's still confusing)

**Comment by brisvag** [comment]
For anyone who might be following: we're working on a separate repo implementing a better version of this: https://github.com/tlambert03/pydantic-generics.

**Comment by brisvag** [comment]
Closing up in favor of #5060.
